From d82841f6c1139d7ef140b8b2f84511a3b1e371b6 Mon Sep 17 00:00:00 2001
From: Patrice Chotard <patrice.chotard@foss.st.com>
Date: Tue, 31 Jan 2023 09:42:41 +0100
Subject: [PATCH 053/120] spi: stm32_ospi: Add calibration algorithm

Calibration algorithm configures the OSPI delay block when OSPI bus
frequency is above STM32_DLYB_FREQ_THRESHOLD (set to 50MHz).

The calibration algorithm steps are the following:
  _ set frequency to low value which doesn't need calibration (50MHz).
  _ send READID command (0x9F) and store response from memory device as
    golden answer.
  _ set frequency to requested value (read from DT).
  _ for each value of Rx tap, send READID command and compare memory
    device answer with golden answer to find Rx Tap valid range.
  _ configure OSPI delay block with Rx tap value located in the middle
    of Rx tap valid range.

If calibration failed (no Rx tap window can be found), the flash
frequency is then set back to a safe value (STM32_DLYB_FREQ_THRESHOLD).

Calibration is always done each time frequency changed on a dedicated
bus.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git e3c089b9227579ca0f57f2b2403877a3d6e5f5ec spi: stm32_ospi: Add calibration algorithm]
Signed-off-by: Patrice Chotard <patrice.chotard@foss.st.com>
Change-Id: I491c39f2416ad1cbbce0a165e5bfab450f9e5b4a
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/288440
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Christophe KERELLO <christophe.kerello@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/308517
---
 drivers/spi/stm32_ospi.c | 259 ++++++++++++++++++++++++++-------------
 1 file changed, 175 insertions(+), 84 deletions(-)

diff --git a/drivers/spi/stm32_ospi.c b/drivers/spi/stm32_ospi.c
index a161fdaf969..7173055a74b 100644
--- a/drivers/spi/stm32_ospi.c
+++ b/drivers/spi/stm32_ospi.c
@@ -17,23 +17,17 @@
 #include <linux/delay.h>
 #include <linux/sizes.h>
 
-struct stm32_ospi_flash {
-	u32 cr;
-	u32 dcr;
-	u32 dcr2;
-	bool initialized;
-};
-
 struct stm32_ospi_priv {
-	struct stm32_ospi_flash flash[OSPI_MAX_CHIP];
 	struct udevice *omi_dev;
 	int cs_used;
+	u64 idcode;
+	u32 prescaler[OSPI_MAX_CHIP];
 };
 
-static int stm32_ospi_mm(struct stm32_ospi_priv *priv,
+static int stm32_ospi_mm(struct udevice *omi_dev,
 			 const struct spi_mem_op *op)
 {
-	struct stm32_omi_plat *omi_plat = dev_get_plat(priv->omi_dev);
+	struct stm32_omi_plat *omi_plat = dev_get_plat(omi_dev);
 
 	memcpy_fromio(op->data.buf.in,
 		      (void __iomem *)omi_plat->mm_base + op->addr.val,
@@ -42,7 +36,7 @@ static int stm32_ospi_mm(struct stm32_ospi_priv *priv,
 	return 0;
 }
 
-static int stm32_ospi_tx(struct stm32_ospi_priv *priv,
+static int stm32_ospi_tx(struct udevice *omi_dev,
 			 const struct spi_mem_op *op,
 			 u8 mode)
 {
@@ -52,14 +46,14 @@ static int stm32_ospi_tx(struct stm32_ospi_priv *priv,
 		return 0;
 
 	if (mode == OSPI_CCR_MEM_MAP)
-		return stm32_ospi_mm(priv, op);
+		return stm32_ospi_mm(omi_dev, op);
 
 	if (op->data.dir == SPI_MEM_DATA_IN)
 		buf = op->data.buf.in;
 	else
 		buf = (u8 *)op->data.buf.out;
 
-	return stm32_omi_tx_poll(priv->omi_dev, buf, op->data.nbytes,
+	return stm32_omi_tx_poll(omi_dev, buf, op->data.nbytes,
 				 op->data.dir == SPI_MEM_DATA_IN);
 }
 
@@ -74,32 +68,22 @@ static int stm32_ospi_get_mode(u8 buswidth)
 	return buswidth;
 }
 
-static int stm32_ospi_exec_op(struct spi_slave *slave,
-			      const struct spi_mem_op *op)
+static int stm32_ospi_send(struct udevice *omi_dev,
+			   const struct spi_mem_op *op, u8 mode)
 {
-	struct stm32_ospi_priv *priv = dev_get_priv(slave->dev->parent);
-	struct stm32_omi_plat *omi_plat = dev_get_plat(priv->omi_dev);
+	struct stm32_omi_plat *omi_plat = dev_get_plat(omi_dev);
+	struct stm32_omi_priv *omi_priv = dev_get_priv(omi_dev);
 	phys_addr_t regs_base = omi_plat->regs_base;
-	u32 cr, ccr = 0, addr_max;
+	u32 cr, ccr = 0;
 	int timeout, ret;
 	int dmode;
-	u8 mode = OSPI_CCR_IND_WRITE;
 	u8 dcyc = 0;
 
-	dev_dbg(slave->dev, "%s: cmd:%#x mode:%d.%d.%d.%d addr:%#llx len:%#x\n",
+	dev_dbg(omi_priv->dev, "%s: cmd:%#x mode:%d.%d.%d.%d addr:%#llx len:%#x\n",
 		__func__, op->cmd.opcode, op->cmd.buswidth, op->addr.buswidth,
 		op->dummy.buswidth, op->data.buswidth,
 		op->addr.val, op->data.nbytes);
 
-	addr_max = op->addr.val + op->data.nbytes + 1;
-
-	if (op->data.dir == SPI_MEM_DATA_IN && op->data.nbytes) {
-		if (addr_max < omi_plat->mm_size && op->addr.buswidth)
-			mode = OSPI_CCR_MEM_MAP;
-		else
-			mode = OSPI_CCR_IND_READ;
-	}
-
 	if (op->data.nbytes)
 		writel(op->data.nbytes - 1, regs_base + OSPI_DLR);
 
@@ -134,7 +118,7 @@ static int stm32_ospi_exec_op(struct spi_slave *slave,
 	if (op->addr.nbytes && mode != OSPI_CCR_MEM_MAP)
 		writel(op->addr.val, regs_base + OSPI_AR);
 
-	ret = stm32_ospi_tx(priv, op, mode);
+	ret = stm32_ospi_tx(omi_dev, op, mode);
 	/*
 	 * Abort in:
 	 * -error case
@@ -146,7 +130,7 @@ static int stm32_ospi_exec_op(struct spi_slave *slave,
 		goto abort;
 
 	/* Wait end of tx in indirect mode */
-	ret = stm32_omi_wait_cmd(priv->omi_dev);
+	ret = stm32_omi_wait_cmd(omi_dev);
 	if (ret)
 		goto abort;
 
@@ -163,21 +147,76 @@ abort:
 	writel(OSPI_FCR_CTCF, regs_base + OSPI_FCR);
 
 	if (ret || timeout)
-		dev_err(slave->dev, "%s ret:%d abort timeout:%d\n", __func__,
+		dev_err(omi_priv->dev, "%s ret:%d abort timeout:%d\n", __func__,
 			ret, timeout);
 
 	return ret;
 }
 
+static int stm32_ospi_exec_op(struct spi_slave *slave,
+			      const struct spi_mem_op *op)
+{
+	struct stm32_ospi_priv *priv = dev_get_priv(slave->dev->parent);
+	struct stm32_omi_plat *omi_plat = dev_get_plat(priv->omi_dev);
+	u32 addr_max;
+	u8 mode = OSPI_CCR_IND_WRITE;
+
+	addr_max = op->addr.val + op->data.nbytes + 1;
+
+	if (op->data.dir == SPI_MEM_DATA_IN && op->data.nbytes) {
+		if (addr_max < omi_plat->mm_size && op->addr.buswidth)
+			mode = OSPI_CCR_MEM_MAP;
+		else
+			mode = OSPI_CCR_IND_READ;
+	}
+
+	return stm32_ospi_send(priv->omi_dev, op, mode);
+}
+
+static int stm32_ospi_readid(struct udevice *omi_dev)
+{
+	struct stm32_omi_priv *omi_priv = dev_get_priv(omi_dev);
+	struct stm32_ospi_priv *ospi_priv = dev_get_priv(omi_priv->dev);
+	u64 rx_buf;
+	struct spi_mem_op readid_op = SPI_MEM_OP(SPI_MEM_OP_CMD(0x9F, 1),
+						 SPI_MEM_OP_NO_ADDR,
+						 SPI_MEM_OP_NO_DUMMY,
+						 SPI_MEM_OP_DATA_IN(8, (u8 *)&rx_buf, 1));
+	int ret;
+
+	ret = stm32_ospi_send(omi_dev, &readid_op, OSPI_CCR_IND_READ);
+	if (ret)
+		return ret;
+	dev_dbg(omi_dev, "Flash ID 0x%08llx\n", rx_buf);
+	/*
+	 * In case of SNAND, the first byte is a dummy byte. Depending of
+	 * memory device, its value can be different in function of frequency.
+	 * Ignore this byte and force its value to 0.
+	 */
+	rx_buf &= 0xffffffffffffff00;
+
+	/* On stm32_ospi_readid() first execution, save the golden READID command's answer */
+	if (ospi_priv->idcode == 0)
+		ospi_priv->idcode = rx_buf;
+
+	if (rx_buf == ospi_priv->idcode)
+		return 0;
+
+	return -EIO;
+}
+
 static int stm32_ospi_probe(struct udevice *bus)
 {
 	struct stm32_ospi_priv *priv = dev_get_priv(bus);
 	struct stm32_omi_plat *omi_plat;
+	struct stm32_omi_priv *omi_priv;
 	phys_addr_t regs_base;
 	int ret;
 
 	priv->omi_dev = bus->parent;
 	omi_plat = dev_get_plat(priv->omi_dev);
+	omi_priv = dev_get_priv(priv->omi_dev);
+	omi_priv->dev = bus;
 	regs_base = omi_plat->regs_base;
 
 	ret = clk_enable(&omi_plat->clk);
@@ -192,63 +231,14 @@ static int stm32_ospi_probe(struct udevice *bus)
 	reset_deassert_bulk(&omi_plat->rst_ctl);
 
 	priv->cs_used = -1;
+	memset(priv->prescaler, -1, OSPI_MAX_CHIP * sizeof(u32));
 
 	setbits_le32(regs_base + OSPI_TCR, OSPI_TCR_SSHIFT);
 
 	/* Set dcr devsize to max address */
-	setbits_le32(regs_base + OSPI_DCR1,
-		     OSPI_DCR1_DEVSIZE_MASK | OSPI_DCR1_DLYBYP);
+	setbits_le32(regs_base + OSPI_DCR1, OSPI_DCR1_DEVSIZE_MASK);
 
-	return 0;
-}
-
-static int stm32_ospi_claim_bus(struct udevice *dev)
-{
-	struct stm32_ospi_priv *priv = dev_get_priv(dev->parent);
-	struct dm_spi_slave_plat *slave_plat = dev_get_parent_plat(dev);
-	struct stm32_omi_plat *omi_plat = dev_get_plat(priv->omi_dev);
-	phys_addr_t regs_base = omi_plat->regs_base;
-	int slave_cs = slave_plat->cs[0];
-
-	if (slave_cs >= OSPI_MAX_CHIP)
-		return -ENODEV;
-
-	if (priv->cs_used != slave_cs) {
-		struct stm32_ospi_flash *flash = &priv->flash[slave_cs];
-
-		priv->cs_used = slave_cs;
-
-		if (flash->initialized) {
-			/* Set the configuration: speed + cs */
-			writel(flash->cr, regs_base + OSPI_CR);
-			writel(flash->dcr, regs_base + OSPI_DCR1);
-			writel(flash->dcr2, regs_base + OSPI_DCR2);
-		} else {
-			/* Set chip select */
-			clrsetbits_le32(regs_base + OSPI_CR,
-					OSPI_CR_CSSEL,
-					priv->cs_used ? OSPI_CR_CSSEL : 0);
-
-			/* Save the configuration: speed + cs */
-			flash->cr = readl(regs_base + OSPI_CR);
-			flash->dcr = readl(regs_base + OSPI_DCR1);
-			flash->dcr2 = readl(regs_base + OSPI_DCR2);
-			flash->initialized = true;
-		}
-	}
-
-	setbits_le32(regs_base + OSPI_CR, OSPI_CR_EN);
-
-	return 0;
-}
-
-static int stm32_ospi_release_bus(struct udevice *dev)
-{
-	struct stm32_ospi_priv *priv = dev_get_priv(dev->parent);
-	struct stm32_omi_plat *omi_plat = dev_get_plat(priv->omi_dev);
-	phys_addr_t regs_base = omi_plat->regs_base;
-
-	clrbits_le32(regs_base + OSPI_CR, OSPI_CR_EN);
+	omi_priv->check_transfer = stm32_ospi_readid;
 
 	return 0;
 }
@@ -261,6 +251,7 @@ static int stm32_ospi_set_speed(struct udevice *bus, uint speed)
 	u32 ospi_clk = omi_plat->clock_rate;
 	u32 prescaler = 255;
 	u32 csht;
+	uint bus_freq;
 	int ret;
 
 	if (speed > 0) {
@@ -284,6 +275,106 @@ static int stm32_ospi_set_speed(struct udevice *bus, uint speed)
 	clrsetbits_le32(regs_base + OSPI_DCR1, OSPI_DCR1_CSHT_MASK,
 			csht << OSPI_DCR1_CSHT_SHIFT);
 
+	bus_freq = ospi_clk / (prescaler + 1);
+
+	if (bus_freq <= STM32_DLYB_FREQ_THRESHOLD)
+		setbits_le32(regs_base + OSPI_DCR1, OSPI_DCR1_DLYBYP);
+	else
+		clrbits_le32(regs_base + OSPI_DCR1, OSPI_DCR1_DLYBYP);
+
+	return 0;
+}
+
+static int stm32_ospi_calibration(struct udevice *bus, uint freq)
+{
+	struct stm32_ospi_priv *priv = dev_get_priv(bus);
+	int ret;
+
+	/*
+	 * Set memory device at low frequency (50MHz) and sent
+	 * READID (0x9F) command, save the answer as golden answer
+	 */
+	ret = stm32_ospi_set_speed(bus, STM32_DLYB_FREQ_THRESHOLD);
+	if (ret)
+		return ret;
+
+	priv->idcode = 0;
+	ret = stm32_ospi_readid(priv->omi_dev);
+	if (ret)
+		return ret;
+
+	/* set frequency at requested value and perform calibration */
+	ret = stm32_ospi_set_speed(bus, freq);
+	if (ret)
+		return ret;
+
+	ret = stm32_omi_dlyb_configure(priv->omi_dev, false, 0);
+	if (ret)
+		return ret;
+
+	/* perform only RX TAP selection */
+	ret = stm32_omi_dlyb_find_tap(priv->omi_dev, true);
+	if (ret)
+		dev_info(bus, "Calibration phase failed\n");
+
+	return ret;
+}
+
+static int stm32_ospi_claim_bus(struct udevice *dev)
+{
+	struct stm32_ospi_priv *priv = dev_get_priv(dev->parent);
+	struct dm_spi_slave_plat *slave_plat = dev_get_parent_plat(dev);
+	struct stm32_omi_plat *omi_plat = dev_get_plat(priv->omi_dev);
+	phys_addr_t regs_base = omi_plat->regs_base;
+	u32 ospi_clk = omi_plat->clock_rate;
+	u32 dcr2, prescaler;
+	int slave_cs = slave_plat->cs[0];
+	uint bus_freq;
+	int ret = 0;
+
+	if (slave_cs >= OSPI_MAX_CHIP)
+		return -ENODEV;
+
+	setbits_le32(regs_base + OSPI_CR, OSPI_CR_EN);
+
+	if (priv->cs_used == slave_cs)
+		return 0;
+
+	priv->cs_used = slave_cs;
+
+	/* Set chip select */
+	clrsetbits_le32(regs_base + OSPI_CR, OSPI_CR_CSSEL,
+			priv->cs_used ? OSPI_CR_CSSEL : 0);
+
+	dcr2 = readl(regs_base + OSPI_DCR2);
+	prescaler = (dcr2 & OSPI_DCR2_PRESC_MASK) >> OSPI_DCR2_PRESC_SHIFT;
+
+	if (prescaler == priv->prescaler[priv->cs_used])
+		return 0;
+
+	priv->prescaler[priv->cs_used] = prescaler;
+	bus_freq = ospi_clk / (prescaler + 1);
+
+	/* calibration needed above 50MHz */
+	if (bus_freq > STM32_DLYB_FREQ_THRESHOLD) {
+		ret = stm32_ospi_calibration(dev->parent, bus_freq);
+		if (ret) {
+			dev_info(dev->parent, "Set flash frequency to a safe value (%d Hz)\n",
+				 STM32_DLYB_FREQ_THRESHOLD);
+			ret = stm32_ospi_set_speed(dev->parent, STM32_DLYB_FREQ_THRESHOLD);
+		}
+	}
+
+	return ret;
+}
+
+static int stm32_ospi_release_bus(struct udevice *dev)
+{
+	struct stm32_ospi_priv *priv = dev_get_priv(dev->parent);
+	struct stm32_omi_plat *omi_plat = dev_get_plat(priv->omi_dev);
+	phys_addr_t regs_base = omi_plat->regs_base;
+
+	clrbits_le32(regs_base + OSPI_CR, OSPI_CR_EN);
 
 	return 0;
 }
-- 
2.47.2

